Cumulative co2 and water

eloch216 opened the pull request:
This PR makes many changes, but the vast majority are related to renaming some parameters and changing some code style, and there are no changes to module behavior (with one small exception; see below). Some tests have been updated, but only because some modules produce new outputs, or because a parameter's name has changed.

The biggest new thing is the addition of two new modules called `BioCro:cumulative_carbon_dynamics` and `BioCro:cumulative_water_dynamics`. These can be optionally added to existing crop models, where they add some cumulative outputs like total transpiration, total photorespiration, etc. These will be handy for determining water use efficiency and for enabling a full "carbon accounting."

This PR is the first of several I am planning to make, which will ultimately include some other new outputs like cumulative growth and maintenance respiration. When it's all done, we will be able to keep track of how much assimilated carbon is used for growth and respiration in each tissue.

Description of changes (copied from `NEWS.md`):
- Renamed several quantities to make their meanings more clear:

  - `mrc1` has been replaced by `grc_leaf` and `grc_stem` to better indicate that this is a growth respiration coefficient that was applied to two different tissues.

  - `mrc2` has been replaced by `grc_root` and `grc_rhizome` for similar reasons.

  - `GrossAssim_CO2` was renamed to `canopy_gross_assimilation_rate_CO2` for consistency with other canopy-level outputs such as `canopy_assimilation_rate_CO2`.

  - `Rd` was renamed to `RL` and its definition was updated when necessary. This is following best practices for naming (to avoid the ambiguity of the subscript `d`, which can refer to "day" or "dark") and an updated understanding of the biochemical origin of this term (it is not exclusively, or even primarily, due to mitochondrial respiration).

  - Several quantities representing a value at 25 degrees C were renamed to better reflect this: `jmax` -> `Jmax_at_25`, `Rd` -> `RL_at_25`, `tpu_rate_max` -> `Tp_at_25`, `vmax1` -> `Vcmax_at_25`, `vmax` -> `Vcmax_at_25`.

- The `BioCro:parameter_calculator` module no longer recalculates `Vcmax_at_25`.

- The rate of non-photorespiratory CO2 release is now included in the outputs of photosynthesis modules, including at the canopy level.

- Two new modules were added for calculating cumulative flows of CO2 and water: `BioCro:cumulative_carbon_dynamics` and `BioCro:cumulative_water_dynamics`.

eloch216 commented:
Here's an example showcasing the new modules:

```R
library(BioCro)

res <- with(soybean, {run_biocro(
    c(initial_values, list(
        canopy_assimilation = 0,
        canopy_gross_assimilation = 0,
        canopy_non_photorespiratory_CO2_release = 0,
        canopy_photorespiration = 0,
        canopy_transpiration = 0,
        soil_evaporation = 0,
        total_precip = 0
    )),
    parameters,
    soybean_weather[['2002']],
    direct_modules,
    c(differential_modules, list(
        'BioCro:cumulative_carbon_dynamics',
        'BioCro:cumulative_water_dynamics'
    ))
)})

dev.new()
xyplot(
    canopy_transpiration + soil_evaporation + total_precip ~ fractional_doy,
    data = res,
    type = 'l',
    auto.key = list(space = 'top'),
    xlab = 'Day of year (2002)',
    ylab = 'Cumulative water flow (Mg / ha)'
)

dev.new()
xyplot(
    canopy_gross_assimilation + canopy_assimilation + canopy_non_photorespiratory_CO2_release + canopy_photorespiration ~ fractional_doy,
    data = res,
    type = 'l',
    auto.key = list(space = 'top'),
    xlab = 'Day of year (2002)',
    ylab = 'Cumulative carbon flow (Mg / ha)'
)
```

Here are the resulting figures:

![image](https://github.com/user-attachments/assets/3a0384a7-31e4-47e3-ad14-df9385e3e38d)

![image](https://github.com/user-attachments/assets/af35f850-afa3-4915-aa01-be0f335d90b0)

yufenghe22 reviewed (approved):
Changing `vmax1` to `Vcmax_at_25` and etc. are long overdue! Thanks for making these changes although my brain has created maps between them.

Just curious, where does the `L` come from in `RL`?
`double RL;                 //!< Rate of non-photorespiratory CO2 release (micromol / m^2 / s)`

eloch216 commented:
> Just curious, where does the `L` come from in `RL`? `double RL; //!< Rate of non-photorespiratory CO2 release (micromol / m^2 / s)`

The `L` comes from "light." I should probably be more consistent about writing that out. Sometimes I feel conflicted about it, because we use that parameter even in the dark.

I am not sure who started calling it `RL` first. I suspect it was Tom Sharkey and/or Berkeley Walker at MSU, since the two earliest papers I know of are https://academic.oup.com/plphys/article/186/1/297/6137851 and https://onlinelibrary.wiley.com/doi/full/10.1111/pce.14153.

I can update this branch to be more clear about the light, and to include a reference for this change.

leavesandlemmas reviewed (approved):
Looks good! The biggest change I would consider is adding the other respiration rates that aren't computed in the canopy to the cumulative carbon dynamics.

The other thing that we might consider adding is the loss of carbon via the litter fluxes. That way it's easy to check if the carbon mass balance is implemented correctly.

However, you might already be planning this in a future PR, so it's not an immediate issue with this PR.

eloch216 commented:
> So, `R_light` and `R_dark` are both mitochondrial respiration. In theory, `R_light` should be smaller than `R_dark` under light conditions due to CO2 re-fixes. In our model however, we use the same `R_light` even under dark conditions. People have been using `Rd` in the FvCB model. Is that wrong because they mean `dark respiration`? The model should use `light respiration`?

In the original 1980 FvCB paper, the authors write:

> The symbol Rd represents CO2 evolution from mitochondria in the light, other than that associated with the PCO cycle.

Here, the "PCO" is "photorespiratory carbon oxidation," so they are saying that `Rd` excludes CO2 released by photorespiration, and they are attributing this non-photorespiratory CO2 release to mitochondrial respiration.

Over the years, variations on this definition have occurred. For example, it is called "non-photorespiratory CO2 release" in https://doi.org/10.1071/pp9840519 and "day respiration other than photorespiration in the light" in https://doi.org/10.1111/pce.12098. You can find many others if you look around.

Eventually, people more clearly began to question the assumption that `Rd` is due to mitochondrial respiration. For example, there is a paper from 2021 (https://doi.org/10.1111/pce.14070) that says:

> Rd has also been called "mitochondrial respiration in the light," but the term "day respiration" is preferred. This is to remain non-specific about where the respired CO2 comes from, as CO2 released is not necessarily mitochondrial in origin.

So I think from the modeling side of things, it's always been very clear that `Rd` is CO2 release occurring in the "light" or "day." Originally it was believed to come from mitochondria, but now it is known that a large portion of it comes from the pentose phosphate pathway (https://doi.org/10.1093/plphys/kiab076).

Experimental papers have made things a bit muddier, because people sometimes estimate `Rd` by measuring the net CO2 assimilation rate in darkness. Strictly speaking, this is not correct, but it's fine as long as people clarify what they're doing. For example, Busch et al. (2020) say the following in the methods:

> We have therefore assumed that Rd in the dark is the same as in the light, and measured Rd in the dark at the beginning of the day of the measurement.

(There is also some other interesting commentary before and after this sentence, justifying the use of Rd estimated this way.)

I think that some people who have mainly worked with measurements and have never taken a close look at the models believe that `Rd` is "respiration in the dark" because it is often estimated that way. This confusion is facilitated by the subscript `d`, which can plausibly refer to "dark." Hence why it is better to use `L`.

In terms of BioCro, I do think it would make sense to different rates of non-photorespiratory CO2 release in the light and the dark. But that's an issue for another day!

eloch216 commented on src/module_library/c4_canopy.h:
```diff
@@ -181,26 +183,27 @@ string_vector c4_canopy::get_inputs()
-        "canopy_conductance",               // mol / m^2 / s
-        "GrossAssim_CO2",                   // micromol CO2 / m^2 / s
-        "canopy_photorespiration_rate_CO2"  // micromol / m^2 / s
+        "canopy_assimilation_rate_CO2",                      // micromol / m^2 / s
```
I had forgotten about that. Justin thought "molar flux" was a clear idea: https://github.com/biocro/biocro/pull/150#issuecomment-2657412786

So I think these suffixes could be changed to `_molar` instead of `_CO2`. That would avoid the current weird name of `canopy_non_photorespiratory_CO2_release_rate_CO2`, which would become `canopy_non_photorespiratory_CO2_release_rate_molar`.

justinmcgrath replied:
canopy_photorespiration_molar_flux, canopy_CO2_assimilation_molar_flux and canopy_non_photorespiratory_CO2_molar_flux are more like what I was thinking of.

eloch216 replied:
Okay, I'll change them to `_molar_flux`.

eloch216 commented on src/module_library/cumulative_carbon_dynamics.h:
```diff
@@ -0,0 +1,93 @@
+        : differential_module{},
+
+          // Get references to input quantities
+          canopy_assimilation_rate{get_input(input_quantities, "canopy_assimilation_rate")},
```
Yes, this was explained in the top comment of this PR:

> This PR is the first of several I am planning to make, which will ultimately include some other new outputs like cumulative growth and maintenance respiration.

eloch216 commented:
I have updated the `RL` documentation and changed the suffixes from `_CO2` to `_molar`. I also noticed that some of the maintenance respiration coefficients were being used for two tissues (despite their names sounding like they only apply to one tissue), so I split them into distinct coefficients.

leavesandlemmas commented:
> > The other thing that we might consider adding is the loss of carbon via the litter fluxes. That way it's easy to check if the carbon mass balance is implemented correctly.
> 
> The litter accumulation is already available in the outputs as `LeafLitter`, `RootLitter`, `StemLitter`, and `RhizomeLitter`.

Okay, so then  if carbon is correctly balanced, then this quantity: 
```cpp
carbon_balance_check =  canopy_gross_assimilation - canopy_respiration_light - canopy_photorespiration -other_respiration_terms  - (LeafLitter + RootLitter + StemLitter + RhizomeLitter) / litter_c_molar_mass  ; 
```
should equal zero at all times.  `litter_c_molar_mass` converts from litter mass to mol carbon.

justinmcgrath commented on src/module_library/parameter_calculator.h:
```diff
@@ -55,30 +50,26 @@ class parameter_calculator : public direct_module
 void parameter_calculator::do_operation() const
 {
     update(lai_op, Leaf * Sp);
-    update(vmax_op, (LeafN_0 - LeafN) * vmax_n_intercept + vmax1);
```
Are you removing this because no one uses it, or is it being done elsewhere. There isn't a clear use now, but this is part of Megan's RIPE objectives.

eloch216 replied:
Previously, this same calculation was made in both `BioCro:parameter_calculator` and `BioCro:linear_vmax_from_leaf_n`. At the moment, pretty much every crop model needs to use `BioCro:parameter_calculator` because it calculates `LAI` from the leaf mass and the specific leaf area. Yet, none of them (currently) use the linear response of Vcmax to leaf nitrogen. I thought it would be simpler to remove this from `BioCro:parameter_calculator` and make it an "optional" approach to Vcmax that can be enabled by adding the `BioCro:linear_vmax_from_leaf_n` module.

Side note: there is also another response to leaf nitrogen inside the multilayer canopy model: https://github.com/biocro/biocro/blob/f8d42d354df2d309d2acc4db72f2fc2a18c4cb27/src/module_library/c3CanAC.cpp#L107-L109. This allows different canopy layers to use different Vcmax.

eloch216 replied:
Since this PR is already a big collection of "changes to improve clarity," I could also divide the remaining parts of the "parameter calculator" into separate modules with clear purpose/name, like `BioCro:lai_from_leaf_mass` and another one for `alpha`.

justinmcgrath replied:
This works. This PR is a big enough as is and I've already looked at it.

yufenghe22 commented:
> Okay, so then if carbon is correctly balanced, then this quantity:
> 
> ```c++
> carbon_balance_check =  canopy_gross_assimilation - canopy_respiration_light - canopy_photorespiration -other_respiration_terms  - (LeafLitter + RootLitter + StemLitter + RhizomeLitter) / litter_c_molar_mass  ; 
> ```
> 
> should equal zero at all times. `litter_c_molar_mass` converts from litter mass to mol carbon.

Shall we add this for diagnosis?

eloch216 commented:
> > Okay, so then if carbon is correctly balanced, then this quantity:
> > ```c++
> > carbon_balance_check =  canopy_gross_assimilation - canopy_respiration_light - canopy_photorespiration -other_respiration_terms  - (LeafLitter + RootLitter + StemLitter + RhizomeLitter) / litter_c_molar_mass  ; 
> > ```
> > should equal zero at all times. `litter_c_molar_mass` converts from litter mass to mol carbon.
> 
> Shall we add this for diagnosis?

I don't think this needs to be a module output; rather, I would just add a test to check that the carbon is balanced. Also, as written it is not quite right, since it doesn't include the mass of the main tissue components (`Leaf`, `Stem`, etc). And all of these will be expressed in `Mg / ha` (or `Mg / ha / hr` for the rates instead of the cumulative totals). So the conversion won't be necessary. This check can be added in a future PR, but right now it is not possible since the maintenance respiration and growth respiration rates / cumulative totals are not yet available as outputs.

justinmcgrath commented:
> So, `R_light` and `R_dark` are both mitochondrial respiration

Ed addressed a lot of this, but I want to point out a couple things. R_light and R_dark are different, but in the context of ACi curves R_d is R_day, not R_dark. R_dark is likely dominated by mitochondrial respiration. However, recent research indicates that CO2 release in the light is mostly from the OPP pathway, and not mitochondrial respiration.

The current thinking is that in the light, very little mitochondrial respiration occurs. I don't fully understand the implications of that. The cell can get NADPH from the OPP pathway, but I don't know where it would get ATP, as I don't think it's exported from the chloroplast.

justinmcgrath commented:
> ...  measured Rd in the dark ...

In other words, they measured respiration in the day in the dark.

leavesandlemmas commented:
> > > Okay, so then if carbon is correctly balanced, then this quantity:
> > > ```c++
> > > carbon_balance_check =  canopy_gross_assimilation - canopy_respiration_light - canopy_photorespiration -other_respiration_terms  - (LeafLitter + RootLitter + StemLitter + RhizomeLitter) / litter_c_molar_mass  ; 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > should equal zero at all times. `litter_c_molar_mass` converts from litter mass to mol carbon.
> > 
> > 
> > Shall we add this for diagnosis?
> 
> I don't think this needs to be a module output; rather, I would just add a test to check that the carbon is balanced. Also, as written it is not quite right, since it doesn't include the mass of the main tissue components (`Leaf`, `Stem`, etc). 

Oops... so we've got
$$
\dot{m} = \mu(A - R) - L  
$$
where $m$ is total mass, $A$ is the whole-plant photosynthetic rate, $R$ is the whole-plant respiration rate, and $L$ is the litter flux. And $\mu$ is the biomass per mol C ratio.  The litter flux conserves the total mass in the plant and in the litter, 

So integrating it should be:
$$
m + m_L = m_0 +  m_{L,0)  +  \int (A - R) dt  
$$
where $m_{L}$ is the mass in the litter.
